File: web/frontend/lib/committees.ts
// Kurznamen für die (langen) Oldenburger Gremiennamen — eine Funktion, überall
// (Design 19a). Amtliche Namen wie „Ausschuss für Wirtschaftsförderung,
// Digitalisierung und internationale Zusammenarbeit" sprengen jede Karte, jedes
// Chip, jedes Dropdown. `shortCommittee` bildet sie auf eine knappe, sinntragende
// Kurzform ab — der volle Name bleibt im title/aria-label und (auf Karten) als
// Unterzeile erhalten. Nie stumpf nach n Zeichen abschneiden.

// Die Oldenburger Gremienliste ist endlich (~15 aktuell + ein paar historische
// Umbenennungen). Gepflegte Tabelle zuerst, Heuristik nur als Fallback.
const SHORT: Record<string, string> = {
  "Rat": "Rat",
  "Rat der Stadt Oldenburg": "Rat",
  "Rat der Stadt Oldenburg (Oldb)": "Rat",
  "Verwaltungsausschuss": "Verwaltungsausschuss",
  "Ausschuss für Allgemeine Angelegenheiten": "Allgemeine Angelegenheiten",
  "Ausschuss für Finanzen und Beteiligungen": "Finanzen & Beteiligungen",
  "Ausschuss für Integration und Migration": "Integration & Migration",
  "Ausschuss für Stadtgrün, Umwelt und Klima": "Stadtgrün & Klima",
  "Ausschuss für Stadtplanung und Bauen": "Stadtplanung & Bauen",
  "Ausschuss für Wirtschaftsförderung, Digitalisierung und internationale Zusammenarbeit": "Wirtschaft & Digitales",
  "Betriebsausschuss Abfallwirtschaftsbetrieb": "Abfallwirtschaft",
  "Betriebsausschuss Eigenbetrieb Gebäudewirtschaft und Hochbau": "Betrieb Gebäudewirtschaft",
  "Jugendhilfeausschuss": "Jugendhilfe",
  "Kulturausschuss": "Kultur",
  "Schulausschuss": "Schule",
  "Sozialausschuss": "Soziales",
  "Sportausschuss": "Sport",
  "Verkehrsausschuss": "Verkehr",
  // historische Umbenennungen (Bestand seit 2018)
  "Ausschuss für Umwelt, Grünflächen und Klimaschutz": "Umwelt & Klima",
  "Ausschuss für Umwelt und Klimaschutz": "Umwelt & Klima",
  "Betriebsausschuss Gebäudewirtschaft und Hochbau": "Betrieb Gebäudewirtschaft",
  "Ausschuss für Wirtschaftsförderung und Digitalisierung": "Wirtschaft & Digitales",
};

/** Amtlicher Gremienname → knappe Kurzform. Unbekannte Namen laufen in eine
 *  Heuristik (Präfix „Ausschuss für …" streichen, „und" → „&", „…ausschuss" →
 *  Kern), nie ein hartes Zeichen-Abschneiden. */
export function shortCommittee(name: string | null | undefined): string {
  if (!name) return "";
  const key = name.trim();
  if (SHORT[key]) return SHORT[key];

  let s = key
    .replace(/^Ausschuss für (den |die |das )?/i, "")
    .replace(/^Betriebsausschuss (Eigenbetrieb )?/i, "");
  // Ein einzelnes „Xausschuss" (Verkehrsausschuss, Kulturausschuss) → Kern.
  if (/^\S+ausschuss$/i.test(s)) s = s.replace(/s?ausschuss$/i, "");
  s = s.replace(/\s+und\s+/gi, " & ");
  s = s.trim();
  return s.length >= 2 ? s : key;
}

/** True, wenn der Kurzname vom vollen Namen abweicht — dann lohnt die Unterzeile. */
export function hasShortCommittee(name: string | null | undefined): boolean {
  return !!name && shortCommittee(name) !== name.trim();
}

// Ein Satz „was hier behandelt wird" je Gremium (Design 26a, Schritt 1). Ohne
// das ist die Abo-Auswahl eine Liste von Amtsbezeichnungen — „Ausschuss für
// Allgemeine Angelegenheiten" sagt niemandem, was drin verhandelt wird.
// Bewusst dieselbe Datei wie shortCommittee: eine Quelle je Gremium.
const EXPLAINS: Record<string, string> = {
  "Rat": "Entscheidet die großen Linien: Haushalt, Satzungen und Grundsatzbeschlüsse.",
  "Verwaltungsausschuss": "Bereitet die Ratsbeschlüsse vor und entscheidet Eilfälle — tagt nichtöffentlich.",
  "Allgemeine Angelegenheiten": "Verwaltung, Personal, Ordnung und alles, was in keinen Fachausschuss fällt.",
  "Finanzen & Beteiligungen": "Haushalt, Zuwendungen und die städtischen Beteiligungen.",
  "Integration & Migration": "Zuwanderung, Teilhabe und interkulturelle Arbeit in der Stadt.",
  "Stadtgrün & Klima": "Grünflächen, Klimaschutz, Energie und Naturschutz in der Stadt.",
  "Umwelt & Klima": "Grünflächen, Klimaschutz, Energie und Naturschutz in der Stadt.",
  "Stadtplanung & Bauen": "Bebauungspläne, Bauprojekte und wie sich Viertel entwickeln.",
  "Wirtschaft & Digitales": "Wirtschaftsförderung, Digitalisierung und internationale Zusammenarbeit.",
  "Abfallwirtschaft": "Müllabfuhr, Recycling und der städtische Abfallbetrieb.",
  "Betrieb Gebäudewirtschaft": "Bau und Unterhalt der städtischen Gebäude — Schulen, Kitas, Verwaltung.",
  "Jugendhilfe": "Kitas, Jugendarbeit und Hilfen für Familien.",
  "Kultur": "Museen, Theater, Bibliotheken und die Förderung der freien Szene.",
  "Schule": "Schulen, Ganztagsbetreuung und neue Bildungsstandorte.",
  "Soziales": "Wohnen, Pflege, Teilhabe und soziale Angebote der Stadt.",
  "Sport": "Sportstätten, Vereinsförderung und Bäder.",
  "Verkehr": "Radwege, Straßen, Bus & Bahn, Parken und Verkehrsberuhigung.",
};

/** Ein Satz, was in diesem Gremium behandelt wird — leer, wenn unbekannt (dann
 *  zeigt die Oberfläche einfach nur den Namen, statt etwas zu erfinden). */
export function committeeExplains(name: string | null | undefined): string {
  if (!name) return "";
  return EXPLAINS[shortCommittee(name)] ?? "";
}

File: ios/Packages/RatslotseFeatures/Sources/RatslotseFeatures/Committees.swift
import RatslotseDesign

/// Was die App über ein Gremium weiß — Kurzname, ein Satz zum Inhalt, sein
/// Zeichen und seine Familie. Spiegel von `web/frontend/lib/committees.ts`,
/// damit beide Frontends dasselbe Zeichen und denselben Kurznamen zeigen;
/// wer dort eine Zeile ergänzt, ergänzt sie hier.
///
/// Die Oldenburger Gremienliste ist endlich (~15 aktuell plus ein paar
/// historische Umbenennungen). Gepflegte Tabelle zuerst, Heuristik nur als
/// Rückfall — nie ein hartes Abschneiden nach n Zeichen.
enum Committee {
    /// Vier Familien, vier Farben (Tims Entscheidung 04.09.2026): Farbe
    /// klassifiziert, statt zu dekorieren — siebzehn Farben wären ein
    /// Regenbogen, vier sagen etwas.
    enum Family: Sendable {
        /// Die Stadt als Ganzes — Rat, Verwaltung, Finanzen, Wirtschaft.
        case city
        /// Das Gebaute — Stadtplanung, Gebäudewirtschaft, Verkehr.
        case built
        /// Grün und Umwelt — Stadtgrün, Klima, Abfall.
        case green
        /// Die Menschen — Soziales, Jugend, Schule, Integration, Kultur, Sport.
        case people

        var accent: RatsWidgetAccent {
            switch self {
            case .city: .harbor
            case .built: .brick
            case .green: .marsh
            case .people: .plum
            }
        }
    }

    struct Entry {
        let short: String
        let explains: String?
        let glyph: RatsGlyph
        let family: Family
    }

    private static let entries: [String: Entry] = {
        func e(_ short: String, _ explains: String?, _ glyph: RatsGlyph, _ family: Family) -> Entry {
            Entry(short: short, explains: explains, glyph: glyph, family: family)
        }
        let council = e("Rat", "Entscheidet die großen Linien: Haushalt, Satzungen und Grundsatzbeschlüsse.", .landmark, .city)
        let green = e("Stadtgrün & Klima", "Grünflächen, Klimaschutz, Energie und Naturschutz in der Stadt.", .leaf, .green)
        let environment = e("Umwelt & Klima", "Grünflächen, Klimaschutz, Energie und Naturschutz in der Stadt.", .leaf, .green)
        let buildings = e("Betrieb Gebäudewirtschaft", "Bau und Unterhalt der städtischen Gebäude — Schulen, Kitas, Verwaltung.", .hammer, .built)
        let economy = e("Wirtschaft & Digitales", "Wirtschaftsförderung, Digitalisierung und internationale Zusammenarbeit.", .laptop, .city)
        return [
            "Rat": council,
            "Rat der Stadt Oldenburg": council,
            "Rat der Stadt Oldenburg (Oldb)": council,
            "Verwaltungsausschuss": e("Verwaltungsausschuss", "Bereitet die Ratsbeschlüsse vor und entscheidet Eilfälle — tagt nichtöffentlich.", .gavel, .city),
            "Ausschuss für Allgemeine Angelegenheiten": e("Allgemeine Angelegenheiten", "Verwaltung, Personal, Ordnung und alles, was in keinen Fachausschuss fällt.", .clipboardList, .city),
            "Ausschuss für Finanzen und Beteiligungen": e("Finanzen & Beteiligungen", "Haushalt, Zuwendungen und die städtischen Beteiligungen.", .coins, .city),
            "Ausschuss für Integration und Migration": e("Integration & Migration", "Zuwanderung, Teilhabe und interkulturelle Arbeit in der Stadt.", .globe, .people),
            "Ausschuss für Stadtgrün, Umwelt und Klima": green,
            "Ausschuss für Stadtplanung und Bauen": e("Stadtplanung & Bauen", "Bebauungspläne, Bauprojekte und wie sich Viertel entwickeln.", .building2, .built),
            "Ausschuss für Wirtschaftsförderung, Digitalisierung und internationale Zusammenarbeit": economy,
            "Betriebsausschuss Abfallwirtschaftsbetrieb": e("Abfallwirtschaft", "Müllabfuhr, Recycling und der städtische Abfallbetrieb.", .recycle, .green),
            "Betriebsausschuss Eigenbetrieb Gebäudewirtschaft und Hochbau": buildings,
            "Jugendhilfeausschuss": e("Jugendhilfe", "Kitas, Jugendarbeit und Hilfen für Familien.", .baby, .people),
            "Kulturausschuss": e("Kultur", "Museen, Theater, Bibliotheken und die Förderung der freien Szene.", .drama, .people),
            "Schulausschuss": e("Schule", "Schulen, Ganztagsbetreuung und neue Bildungsstandorte.", .graduationCap, .people),
            "Sozialausschuss": e("Soziales", "Wohnen, Pflege, Teilhabe und soziale Angebote der Stadt.", .heartHandshake, .people),
            "Sportausschuss": e("Sport", "Sportstätten, Vereinsförderung und Bäder.", .quiz, .people),
            "Verkehrsausschuss": e("Verkehr", "Radwege, Straßen, Bus & Bahn, Parken und Verkehrsberuhigung.", .bus, .built),
            // historische Umbenennungen (Bestand seit 2018)
            "Ausschuss für Umwelt, Grünflächen und Klimaschutz": environment,
            "Ausschuss für Umwelt und Klimaschutz": environment,
            "Betriebsausschuss Gebäudewirtschaft und Hochbau": buildings,
            "Ausschuss für Wirtschaftsförderung und Digitalisierung": economy,
            "Ausschuss für Wirtschaftsförderung und internationale Zusammenarbeit": economy,
            "Ausschuss für Bahnangelegenheiten": e("Bahnangelegenheiten", nil, .bus, .built),
        ]
    }()

    /// Der gepflegte Eintrag — oder ein aus dem Namen gebauter: Kurzname per
    /// Heuristik, Gruppe als Zeichen (ein Gremium sind Menschen, die
    /// zusammensitzen; das ist nie falsch, nur allgemein), Familie „Stadt".
    static func entry(_ name: String) -> Entry {
        let key = name.trimmingCharacters(in: .whitespacesAndNewlines)
        if let known = entries[key] { return known }
        return Entry(short: heuristicShort(key), explains: nil, glyph: .users, family: .city)
    }

    static func short(_ name: String) -> String { entry(name).short }
    static func glyph(_ name: String) -> RatsGlyph { entry(name).glyph }
    static func accent(_ name: String) -> RatsWidgetAccent { entry(name).family.accent }

    /// Der Rat selbst — nicht Ortsrat, Beirat oder Verwaltungsausschuss.
    static func isCouncil(_ name: String) -> Bool {
        RatslotseFeatures.isCouncil(name)
    }

    /// Präfix „Ausschuss für …" streichen, „und" → „&", „…ausschuss" → Kern —
    /// dieselben Regeln wie `shortCommittee` im Web.
    private static func heuristicShort(_ key: String) -> String {
        var s = key
        for prefix in ["Ausschuss für den ", "Ausschuss für die ", "Ausschuss für das ", "Ausschuss für ",
                       "Betriebsausschuss Eigenbetrieb ", "Betriebsausschuss "] where s.hasPrefix(prefix) {
            s.removeFirst(prefix.count)
            break
        }
        if !s.contains(" "), s.lowercased().hasSuffix("ausschuss") {
            s.removeLast("ausschuss".count)
            if s.hasSuffix("s") { s.removeLast() }
        }
        s = s.replacingOccurrences(of: " und ", with: " & ").trimmingCharacters(in: .whitespaces)
        return s.count >= 2 ? s : key
    }
}
